### PR TITLE
[AArch64] Avoid single-element vector fp converts in streaming[-compatible] functions

### PR DESCRIPTION
**Pull request by MacDue**
The single-element vector variants of FCVTZS, FCVTZU, UCVTF, and SCVTF are only supported in streaming[-compatible] functions with `+sme2p2`.

Reference:
- https://developer.arm.com/documentation/ddi0602/2024-09/SIMD-FP-Instructions/FCVTZS--vector--integer---Floating-point-convert-to-signed-integer--rounding-toward-zero--vector--
- https://developer.arm.com/documentation/ddi0602/2024-09/SIMD-FP-Instructions/UCVTF--vector--integer---Unsigned-integer-convert-to-floating-point--vector--
- https://developer.arm.com/documentation/ddi0602/2024-09/SIMD-FP-Instructions/SCVTF--vector--integer---Signed-integer-convert-to-floating-point--vector--

Codegen will be improved in follow up patches.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 1]
```diff
@@ -247,6 +247,11 @@ def HasSMEF16F16orSMEF8F16
 def HasNEONandIsStreamingSafe
     : Predicate<"Subtarget->hasNEON()">,
       AssemblerPredicateWithAll<(any_of FeatureNEON), "neon">;
+// A subset of NEON instructions legal in Streaming SVE mode with +sme2p2.
+// TODO: Change to check for hasSME2p2() once FEAT_SME2p2 is implemented.
+def HasNEONandIsSME2p2StreamingSafe
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:252]: nit: At the moment the predicate does not do what the name suggests. Could you use the existing `HasNEON` predicate and instead add a TODO to the places where `HasNEONandIsStreamingSafe` was used incorrectly?

File: llvm/test/CodeGen/AArch64/sve-streaming-mode-cvt-fp-int-fp.ll [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+; RUN: llc -force-streaming-compatible  < %s | FileCheck %s
+; RUN: llc < %s | FileCheck %s --check-prefix=NON-STREAMING
+
+target triple = "aarch64-unknown-linux-gnu"
+
+define double @t1(double %x) {
+; CHECK-LABEL: t1:
+; CHECK:       // %bb.0: // %entry
+; CHECK-NEXT:    fcvtzs x8, d0
+; CHECK-NEXT:    scvtf d0, x8
```
Comment (sdesmalen-arm) [inline llvm/test/CodeGen/AArch64/sve-streaming-mode-cvt-fp-int-fp.ll:11]: This operation may be expensive when executed in streaming mode, because the operations move results between register files (FPR <-> GPR). Could you emulate this operation with SVE instructions instead?
I'm happy for this to be done in a separate patch so that this bugfix can already land.

Comment (MacDue) [reply llvm/test/CodeGen/AArch64/sve-streaming-mode-cvt-fp-int-fp.ll:11]: I'll look into that in a follow up patch :+1: I need to check how to do something ~equivalent for SVE since for Neon this are a load of .td patterns (matching both the to/from int parts).

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 3]
```diff
@@ -6237,7 +6237,8 @@ def : Pat<(v2f64 (AArch64frsqrts (v2f64 FPR128:$Rn), (v2f64 FPR128:$Rm))),
 // Some float -> int -> float conversion patterns for which we want to keep the
 // int values in FP registers using the corresponding NEON instructions to
 // avoid more costly int <-> fp register transfers.
-let Predicates = [HasNEONandIsStreamingSafe] in {
+// TODO: Allow these in streaming[-compatible] functions with +sme2p2.
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:6240]: nit: could you replicate this comment on the other two uses too?

**Review by sdesmalen-arm — APPROVED** [review]
